bitcoin/src/blockdata/script/txhash.rs

```diff
@@ -0,0 +1,1148 @@
+                    return Err("relative individual output selected out of range");
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+fn offset(relative: &[isize], current_input_idx: usize) -> Result<Vec<usize>, String> {
+    relative.iter().map(|i| {
+        (current_input_idx as isize).checked_add(*i)
+            .ok_or("relative index overflow")?
+            .try_into().map_err(|_| "relative index underflow".to_owned())
+    }).collect()
+}
+
+struct InputsCache {
+    scriptsigs: Vec<Option<sha256::Hash>>,
+    prev_spks: Vec<Option<sha256::Hash>>,
+    annexes: Vec<Option<sha256::Hash>>,
+    control_blocks: Vec<Option<sha256::Hash>>,
+    spentscripts: Vec<Option<sha256::Hash>>,
+}
+
+impl InputsCache {
+    fn new(nb_inputs: usize) -> InputsCache {
+        InputsCache {
+            scriptsigs: Vec::new_default(nb_inputs),
+            prev_spks: Vec::new_default(nb_inputs),
+            annexes: Vec::new_default(nb_inputs),
+            control_blocks: Vec::new_default(nb_inputs),
+            spentscripts: Vec::new_default(nb_inputs),
+        }
+    }
+
+    fn scriptsig(&mut self, idx: usize, tx: &Transaction) -> sha256::Hash {
+        *self.scriptsigs[idx].get_or_insert_with(|| {
+            if tx.input[idx].script_sig.is_empty() {
+                SHA256_EMPTY
+            } else {
+                sha256::Hash::hash(tx.input[idx].script_sig.as_bytes())
+            }
+        })
+    }
+
+    fn prev_spk(&mut self, idx: usize, prevouts: &[TxOut]) -> sha256::Hash {
+        *self.prev_spks[idx].get_or_insert_with(|| {
+            sha256::Hash::hash(prevouts[idx].script_pubkey.as_bytes())
+        })
+    }
+
+    fn annex(&mut self, idx: usize, tx: &Transaction, prevouts: &[TxOut]) -> sha256::Hash {
+        *self.annexes[idx].get_or_insert_with(|| {
+            if prevouts[idx].script_pubkey.is_p2tr() {
+                // NB this will only be called for inputs that are p2tr
+                if let Some(annex) = tx.input[idx].witness.taproot_annex() {
+                    sha256::Hash::hash(annex)
+                } else {
+                    SHA256_EMPTY
+                }
+            } else {
+                SHA256_EMPTY
+            }
+        })
+    }
+
+    fn control_block(&mut self, idx: usize, tx: &Transaction) -> sha256::Hash {
+        *self.control_blocks[idx].get_or_insert_with(|| {
+            // NB this will only be called for inputs that are p2tr
+            if let Some(cb) = tx.input[idx].witness.taproot_control_block() {
+                sha256::Hash::hash(cb)
+            } else {
+                SHA256_EMPTY
+            }
+        })
+    }
+
+    fn spentscript(&mut self, idx: usize, tx: &Transaction, prev_script: &Script) -> sha256::Hash {
+        // NB this will only be called for inputs that are p2sh, p2wsh or p2tr
+        *self.spentscripts[idx].get_or_insert_with(|| {
+            if prev_script.is_p2sh() {
```

Comment:
I've been looking at a way to find the type of script in a different issue https://github.com/rust-bitcoin/rust-bitcoin/issues/2609.  Seems like this this could be consolidated by matching on an enum of script type.